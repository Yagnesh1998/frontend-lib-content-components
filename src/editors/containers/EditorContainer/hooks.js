import { useSelector } from 'react-redux';

import { RequestKeys } from '../../data/constants/requests';
import { selectors } from '../../data/redux';
import * as appHooks from '../../hooks';
import analyticsEvt from '../../data/constants/analyticsEvt';

export const {
  navigateCallback,
  nullMethod,
  saveBlock,
} = appHooks;

export const setAssetToStaticUrl = (images, getContent) => {
  let content = getContent();
  const imageUrls = [];
  const imgsArray = Object.values(images);
  imgsArray.forEach(image => {
    imageUrls.push({ portableUrl: image.portableUrl, displayName: image.displayName });
  });
  const imageSrcs = content.split('src="');
  imageSrcs.forEach(src => {
    if (src.startsWith('/asset') && imageUrls.length > 0) {
      const nameFromEditorSrc = src.substring(src.lastIndexOf('@') + 1, src.indexOf('"'));
      const nameFromStudioSrc = nameFromEditorSrc.substring(nameFromEditorSrc.indexOf('/') + 1);
      let portableUrl;
      imageUrls.forEach((url) => {
        if (url.displayName === nameFromEditorSrc || url.displayName === nameFromStudioSrc) {
          portableUrl = url.portableUrl;
        }
      });
      if (portableUrl) {
        const currentSrc = src.substring(0, src.indexOf('"'));
        const updatedContent = content.replace(currentSrc, portableUrl);
        content = updatedContent;
      }
    }
  });
  return content;
};

export const handleSaveClicked = ({ getContent, dispatch }) => {
  const destination = useSelector(selectors.app.returnUrl);
  const analytics = useSelector(selectors.app.analytics);
  const images = useSelector(selectors.app.images);
  return () => saveBlock({
    content: setAssetToStaticUrl(images, getContent),
    destination,
    analytics,
    dispatch,
  });
};
export const handleCancelClicked = ({ onClose }) => {
  if (onClose) {
    return onClose;
  }
  return navigateCallback({
    destination: useSelector(selectors.app.returnUrl),
    analyticsEvent: analyticsEvt.editorCancelClick,
    analytics: useSelector(selectors.app.analytics),
  });
};
export const isInitialized = () => useSelector(selectors.app.isInitialized);
export const saveFailed = () => useSelector((state) => (
  selectors.requests.isFailed(state, { requestKey: RequestKeys.saveBlock })
));
