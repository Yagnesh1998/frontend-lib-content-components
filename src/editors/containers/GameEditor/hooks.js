import { useState, useEffect } from 'react';
import { Form } from '@edx/paragon';
import './index.scss';
