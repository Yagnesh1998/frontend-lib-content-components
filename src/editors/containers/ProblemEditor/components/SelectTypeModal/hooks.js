import {
  useState,
} from 'react';
import { StrictDict } from '../../../../utils';

export const state = StrictDict({
  selected: (val) => useState(val),
});

export const onSelect = ({ selected }) => {
  return;
};

export default { state, onSelect };
