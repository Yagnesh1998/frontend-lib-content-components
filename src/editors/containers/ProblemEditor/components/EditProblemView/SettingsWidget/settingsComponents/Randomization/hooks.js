import { useState, useEffect } from 'react';
import messages from './messages';
import { RandomizationTypes } from '../../../../../../../data/constants/problem';
import * as module from './hooks';

export const state = {
  summary: (val) => useState(val),
};

export const useRandomizationSettingStatus = ({ randomization, updateSettings }) => {
  const [summary, setSummary] = module.state.summary({ message: messages.noRandomizationSummary, values: {} });
  useEffect(() => {
    setSummary({
      message: randomization ? RandomizationTypes[randomization] : messages.noRandomizationSummary,
    });
  }, [randomization]);

  const handleChange = (event) => {
    updateSettings({ randomization: event.target.value });
  };
  return { summary, handleChange };
};

export default useRandomizationSettingStatus;
