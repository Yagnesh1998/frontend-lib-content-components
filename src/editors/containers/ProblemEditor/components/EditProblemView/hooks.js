import 'tinymce';
import ReactStateSettingsParser from '../../data/ReactStateSettingsParser';
import ReactStateOLXParser from '../../data/ReactStateOLXParser';
import { setAssetToStaticUrl } from '../../../../sharedComponents/TinyMceWidget/hooks';

export const fetchEditorContent = () => {
  const editorObject = {};
  const EditorsArray = window.tinymce.editors;
  Object.entries(EditorsArray).forEach(([id, editor]) => {
    if (Number.isNaN(parseInt(id))) {
      if (id.startsWith('answer')) {
        const { answers } = editorObject;
        const answerId = id.substring(id.indexOf('-') + 1);
        editorObject.answers = { ...answers, [answerId]: editor.getContent() };
      } else {
        editorObject[id] = editor.getContent();
      }
    }
  });
  return editorObject;
};

export const parseState = ({
  problem,
  isAdvanced,
  ref,
  assets,
  lmsEndpointUrl,
}) => () => {
  const editorObject = fetchEditorContent();
  const reactSettingsParser = new ReactStateSettingsParser(problem);
  const reactOLXParser = new ReactStateOLXParser({ problem, editorObject });
  const reactBuiltOlx = setAssetToStaticUrl({ editorValue: reactOLXParser.buildOLX(), assets, lmsEndpointUrl });
  const rawOLX = ref?.current?.state.doc.toString();
  return {
    settings: reactSettingsParser.getSettings(),
    olx: isAdvanced ? rawOLX : reactBuiltOlx,
  };
};
