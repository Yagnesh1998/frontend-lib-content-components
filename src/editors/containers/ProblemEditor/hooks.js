import {
  useRef, useEffect, useCallback, useState,
} from 'react';
import { useSelector } from 'react-redux';
import { selectors } from '../../data/redux/app';
import { StrictDict } from '../../utils';

import { RequestKeys } from '../../data/constants/requests';

import * as module from './hooks';

export const state = StrictDict({
  selected: (val) => useState(val),

});

export const getStep = () => {
  const problemType = useSelector(selectors.problem.problemType);
  return {
    problemType,
  };
};
