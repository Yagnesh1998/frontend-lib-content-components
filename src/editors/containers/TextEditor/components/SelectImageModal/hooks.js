import React from 'react';
import { useDispatch } from 'react-redux';

import { thunkActions } from '../../../../data/redux';
import * as module from './hooks';
import { sortFunctions, sortKeys } from './utils';

export const state = {
  highlighted: (val) => React.useState(val),
  images: (val) => React.useState(val),
  showSelectImageError: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
  showSizeError: (val) => React.useState(val),
};

export const searchAndSortHooks = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  return {
    searchString,
    onSearchChange: (e) => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: (key) => () => setSortBy(key),
  };
};

export const filteredList = ({ searchString, imageList }) => (
  imageList.filter(({ displayName }) =>displayName.toLowerCase().includes(searchString.toLowerCase()))
);

export const displayList = ({ sortBy, searchString, images }) =>
  module.filteredList({
      searchString,
      imageList: Object.values(images),
    }).sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);

export const imgListHooks = ({ searchSortProps, setSelection }) => {
  const dispatch = useDispatch();
  const [images, setImages] = module.state.images({});
  const [highlighted, setHighlighted] = module.state.highlighted(null);
  const [
    showSelectImageError,
    setShowSelectImageError,
  ] = module.state.showSelectImageError(false);
  const [showSizeError, setShowSizeError] = module.state.showSelectImageError(false);
  const list = module.displayList({ ...searchSortProps, images });

  React.useEffect(() => {
    dispatch(thunkActions.app.fetchImages({ setImages }));
  }, []);

  return {
    galleryError: {
      show: showSelectImageError,
      set: () => setShowSelectImageError(true),
      dismiss: () => setShowSelectImageError(false),
    },
    inputError: {
      show: showSizeError,
      set: () => setShowSizeError(true),
      dismiss: () => setShowSizeError(false),
    },
    images,
    galleryProps: {
      galleryIsEmpty: Object.keys(images).length === 0,
      searchIsEmpty: list.length === 0,
      displayList: list,
      highlighted,
      onHighlightChange: (e) => setHighlighted(e.target.value),
    },
    // highlight by id
    selectBtnProps: {
      onClick: () => {
        if (highlighted) {
          setSelection(images[highlighted]);
        } else {
          setShowSelectImageError(true);
        }
      },
    },
  };
};

export const checkValidFileSize = ({
  selectedFile,
  clearSelection,
  onSizeFail,
}) => {
  const fileSize = (selectedFile.size / 1000000).toFixed(4);
  if (fileSize > 10) {
    clearSelection();
    onSizeFail();
    return false;
  }
  return true;
};

export const fileInputHooks = ({ setSelection, clearSelection, imgList }) => {
  const dispatch = useDispatch();
  const ref = React.useRef();
  const click = () => ref.current.click();
  const addFile = (e) => {
    const selectedFile = e.target.files[0];
    if (selectedFile && module.checkValidFileSize({
        selectedFile,
        clearSelection,
        onSizeFail: () => {
          imgList.inputError.set();
        },
      })
    ) {
      dispatch(
        thunkActions.app.uploadImage({
          file: selectedFile,
          setSelection,
        }),
      );
    }
  };

  return {
    click,
    addFile,
    ref,
  };
};

export const imgHooks = ({ setSelection, clearSelection }) => {
  const searchSortProps = module.searchAndSortHooks();
  const imgList = module.imgListHooks({ setSelection, searchSortProps });
  const fileInput = module.fileInputHooks({
    setSelection,
    clearSelection,
    imgList,
  });
  const {
    galleryError,
    galleryProps,
    inputError,
    selectBtnProps,
  } = imgList;

  return {
    galleryError,
    inputError,
    fileInput,
    galleryProps,
    searchSortProps,
    selectBtnProps,
  };
};

export default {
  imgHooks,
};
