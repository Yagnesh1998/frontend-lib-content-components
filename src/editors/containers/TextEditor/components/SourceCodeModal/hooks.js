import { useRef } from 'react';
import * as module from './hooks';

export const getSaveBtnProps = ({ editorRef, ref, close }) => ({
  onClick: () => {
    if (editorRef && editorRef.current && ref && ref.current) {
      const content = ref.current.state.doc.toString();
      editorRef.current.setContent(content);
      close();
    }
  },
});

export const preprareSourceCodeModal = ({ editorRef, close }) => {
  const ref = useRef();
  const saveBtnProps = module.getSaveBtnProps({ editorRef, ref, close });
  const value = editorRef.current?.getContent();
  return { saveBtnProps, value, ref };
};

export default {
  preprareSourceCodeModal,
};
