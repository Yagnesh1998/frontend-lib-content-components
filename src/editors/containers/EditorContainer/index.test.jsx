import { IconButton } from '@edx/paragon';
import { shallow } from 'enzyme';
import { useDispatch } from 'react-redux';

import { EditorContainer } from '.';
import * as hooks from './hooks';

const props = {
  getContent: jest.fn().mockName('props.getContent'),
  onClose: jest.fn().mockName('props.onClose'),
  validateEntry: jest.fn().mockName('props.validateEntry'),
};

jest.mock('./hooks', () => ({
  isInitialized: jest.fn().mockReturnValue(true),
  handleCancelClicked: (args) => ({ handleCancelClicked: args }),
  handleSaveClicked: (args) => ({ handleSaveClicked: args }),
  saveFailed: jest.fn().mockName('hooks.saveFailed'),
}));

let el;

describe('EditorContainer component', () => {
  describe('render', () => {
    const testContent = (<h1>My test content</h1>);
    test('snapshot: not initialized. disable save and pass to header', () => {
      hooks.isInitialized.mockReturnValueOnce(false);
      expect(
        shallow(<EditorContainer {...props}>{testContent}</EditorContainer>),
      ).toMatchSnapshot();
    });
    test('snapshot: initialized. enable save and pass to header', () => {
      expect(
        shallow(<EditorContainer {...props}>{testContent}</EditorContainer>),
      ).toMatchSnapshot();
    });
    describe('behavior inspection', () => {
      beforeEach(() => {
        el = shallow(<EditorContainer {...props}>{testContent}</EditorContainer>);
      });

      test('close behavior is linked to modal onClose', () => {
        const expected = hooks.handleCancelClicked({ onClose: props.onClose });
        expect(el.find(IconButton)
          .props().onClick).toEqual(expected);
      });
      test('close behavior is linked to footer onCancel', () => {
        const expected = hooks.handleCancelClicked({ onClose: props.onClose });
        expect(el.children().at(2)
          .props().onCancel).toEqual(expected);
      });
      test('save behavior is linked to footer onSave', () => {
        const expected = hooks.handleSaveClicked({
          dispatch: useDispatch(),
          getContent: props.getContent,
          validateEntry: props.validateEntry,
        });
        expect(el.children().at(2)
          .props().onSave).toEqual(expected);
      });
    });
  });
});
