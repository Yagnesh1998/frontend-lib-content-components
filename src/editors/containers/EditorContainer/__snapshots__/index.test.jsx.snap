// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorContainer component render snapshot: initialized. enable save and pass to header 1`] = `
<div>
  <ModalDialog.Header
    className="shadow-sm zindex-10"
  >
    <ModalDialog.Title>
      <div
        style={
          Object {
            "height": "44px",
            "margin": "auto",
          }
        }
      >
        <injectIntl(ShimmedIntlComponent)
          isInitialized={true}
        />
      </div>
      <div
        className="pgn__modal-close-container"
      >
        <IconButton
          iconAs="Icon"
          onClick={
            Object {
              "handleCancelClicked": Object {
                "onClose": [MockFunction props.onClose],
              },
            }
          }
          src={[MockFunction icons.Close]}
        />
      </div>
    </ModalDialog.Title>
  </ModalDialog.Header>
  <h1>
    My test content
  </h1>
  <injectIntl(ShimmedIntlComponent)
    disableSave={false}
    onCancel={
      Object {
        "handleCancelClicked": Object {
          "onClose": [MockFunction props.onClose],
        },
      }
    }
    onSave={
      Object {
        "handleSaveClicked": Object {
          "dispatch": [MockFunction react-redux.dispatch],
          "getContent": [MockFunction props.getContent],
          "validateEntry": [MockFunction props.validateEntry],
        },
      }
    }
  />
</div>
`;

exports[`EditorContainer component render snapshot: not initialized. disable save and pass to header 1`] = `
<div>
  <ModalDialog.Header
    className="shadow-sm zindex-10"
  >
    <ModalDialog.Title>
      <div
        style={
          Object {
            "height": "44px",
            "margin": "auto",
          }
        }
      >
        <injectIntl(ShimmedIntlComponent)
          isInitialized={false}
        />
      </div>
      <div
        className="pgn__modal-close-container"
      >
        <IconButton
          iconAs="Icon"
          onClick={
            Object {
              "handleCancelClicked": Object {
                "onClose": [MockFunction props.onClose],
              },
            }
          }
          src={[MockFunction icons.Close]}
        />
      </div>
    </ModalDialog.Title>
  </ModalDialog.Header>
  <injectIntl(ShimmedIntlComponent)
    disableSave={true}
    onCancel={
      Object {
        "handleCancelClicked": Object {
          "onClose": [MockFunction props.onClose],
        },
      }
    }
    onSave={
      Object {
        "handleSaveClicked": Object {
          "dispatch": [MockFunction react-redux.dispatch],
          "getContent": [MockFunction props.getContent],
          "validateEntry": [MockFunction props.validateEntry],
        },
      }
    }
  />
</div>
`;
